Added a nice way to create tables

pb82 opened the pull request:
I thought it would be useful to have a nice way of creating tables, so I added a function to do this. The table definition is passed as a keyword list, like:

```elixir
Sqlitex.create(db, "users", id: :integer, name: :text)
```

Since sqlite does not have types, i don't validate them. Sqlite literally allows you to pass anything as column type and it will choose an affinity based on some rules (https://www.sqlite.org/datatype3.html). Column names and types are only required to be passed as atoms.

I've also added a test and extended the server api. Let me know what you think,

Thanks!

mmmries commented:
@pb82 thanks for contributing to the project!

What do you think of calling the new method `create_table`? It probably isn't strictly necessary since there isn't much else that you could create in a database, but I like the idea of having the Sqlitex API match the SQL keywords wherever it is convenient to do so.

I'm curious to explore this API a bit more before we try to merge it in and release it. Looking at the [create table docs](https://www.sqlite.org/lang_createtable.html) it looks like there are several optional settings for tables like `TEMPORARY` or `IF NOT EXISTS`. There are also column-level optional settings like `PRIMARY KEY`, `NOT NULL` [and other similar column definition flags](https://www.sqlite.org/syntax/column-constraint.html). I don't think we need to support all of these flags on the first pass, but I think we should explore supporting a few of these settings at both the `table` and `column` levels so that we don't get into a situation where we have to make a backwards incompatible API change to Sqlitex.

If you are open to exploring those ideas I'm happy to jump in and help or stay hands off and let you explore.

pb82 commented:
You are right `create_table` is more appropriate. Just `create` might be confused with `create_database`. Will change this. As for the table and column options: i think some of them can be supported in a reasonable way (the ones that don't take expressions). At the moment i'm thinking of a good way to design the API (probably with 3-tuples for colum definitions) . I'll update the pull request as soon as possible.

Thanks for the review!

jazzyb commented:
Just my two cents:  I don't know that all the CREATE TABLE options are necessary, but [optional column constraints](https://www.sqlite.org/syntax/column-constraint.html) would be useful since, unlike declared types, SQLite actually validates values based on these constraints.

pb82 commented:
Hey,

i've updated the pull request with a new commit. Renamed to `create_table`. Some table and column constraints are now supported and there is a new module: SqlBuilder. SqlBuilder can easily be extended to support new constraints.

For now the supported table constraints are: TEMPORARY and PRIMARY KEY (...)
The supported column constraints are: NOT NULL, PRIMARY KEY and AUTOINCREMENT

The API for create_table would look like this:

```elixir
Sqlitex.create_table(db, 
  :users,   # table name
  [:temp],  # (optional) table constaints
  id: {:integer, [:primary_key]}, name: :text) # columns with and without constraints
```

I've added some documentation and test for the new functionality.

mmmries commented:
@pb82 intially I think it looks really cool. Thanks again for spending your time making this library more awesome! :heart: :heart_eyes: :zap:

I'll hopefully have some more time this evening to pull down the branch and play around with it a bit.

jazzyb commented on lib/sqlitex/server.ex:
```diff
@@ -43,6 +48,10 @@ defmodule Sqlitex.Server do
   end
 
+  def create_table(pid, name, cols) do
+    GenServer.call(pid, {:create_table, name, cols})
```
The `Sqlitex.create_table` function takes table options.  Should this server function take the same optional table options?

jazzyb commented on lib/sqlitex/sql_builder.ex:
```diff
@@ -0,0 +1,74 @@
+defmodule Sqlitex.SqlBuilder do
```
I love this SqlBuilder module.  Very nice work.  I regret that I won't get a chance to test it until later this week.  We should definitely think about expanding this in the future for other SQL commands, but that doesn't need to happen before this merge.

mmmries replied:
I don't know much about `Ecto`, but I know they have spent a lot of time on their API for creating SQL queries. I wonder how much value there is in making our own SQL generator vs doing the SQL creation at the Ecto layer?

jazzyb replied:
Every Ecto adapter (MySQL, MS SQL, PostgreSQL, and SQLite) has had to implement their own SQL query generation due to differing features and functionality in each respective SQL dialect -- not that any of it is *significantly* different, rather they all fail to overlap just enough that, so far, there is little code reuse.  I like this module because the sort of SQL generation I have been writing for sqlite_ecto could be shifted to this module in the distant future.

Having said that, whether or not you want to merge this module comes down to the direction you would like Sqlitex to go:  It could remain a "dumb" (I don't use that term derogatorily) wrapper around esqlite, or it could become a sort-of-DSL for SQLite databases.  I think there is value in either direction.

jazzyb commented on lib/sqlitex/sql_builder.ex:
```diff
@@ -0,0 +1,74 @@
+  # passed keyword list
+  defp get_columns_block(cols) do
+    Enum.reduce(cols, "", fn(col, acc) ->
+      comma = if acc == "" do nil else "," end
```
Correct me if I'm wrong, but it looks like you are trying to account for the comma separators between column definitions.  It might be cleaner and easier to `map` the columns to a list of individual definitions and then pipe that output to `Enum.join(", ")`.  For example:

```elixir
cols
|> Enum.map(fn
  {name, {type, constraints}} -> # blah
  {name, type} -> # blah
end)
|> Enum.join(", ")
```

The above code can be further simplified with [`Enum.map_join/3`](http://elixir-lang.org/docs/stable/elixir/Enum.html#map_join/3).

jazzyb commented on lib/sqlitex/sql_builder.ex:
```diff
@@ -0,0 +1,74 @@
+    tbl_options = get_opts_dict(table_opts, &table_opt/1)
+    get_opt = &(Dict.get(tbl_options, &1, nil))
+
+    "CREATE #{get_opt.(:temp)} TABLE #{get_opt.(:if_not_exists)} #{name} (#{get_columns_block(cols)} #{get_opt.(:primary_key)})"
```
Look into how SQLite quotes identifiers such as table names.  Quoting identifiers allows users to define tables and columns using otherwise illegal characters.  For example:

```
sqlite> create table t-1 (a, b, c);
Error: near "-": syntax error
sqlite> create table "t-1" (a, b, c);
sqlite> .schema
CREATE TABLE "t-1" (a, b, c);
sqlite>
```

Ditto for column names in `get_columns_block/1` above.

jazzyb commented on lib/sqlitex/sql_builder.ex:
```diff
@@ -0,0 +1,74 @@
+  end
+
+  # Catch invalid / unsupported table options
+  defp table_opt(opt), do: raise "Unkown table option: #{IO.inspect(opt)}"
```
You don't need to raise an error here.  Unmatched arguments will raise runtime exceptions anyway.

jazzyb commented on lib/sqlitex/sql_builder.ex:
```diff
@@ -0,0 +1,74 @@
+  defp column_opt(:autoincrement), do: {:autoincrement, "AUTOINCREMENT"}
+
+  # Catch invalid / unsupported column options
+  defp column_opt(opt), do: raise "Unknown column option: #{IO.inspect(opt)}"
```
Ditto.

jazzyb commented on test/sql_builder_test.exs:
```diff
@@ -0,0 +1,19 @@
+
+    sql = Sql.create_table(:users, [], id: {:integer, [:not_null, :primary_key]}, name: :text)
+
+    assert sql == "CREATE  TABLE  users ( id integer PRIMARY KEY NOT NULL , name text )"
```
Can you write another assertion for the `PRIMARY KEY` table constraint?

mmmries commented:
This looks really good. @pb82 thanks for sending in this pull request and for putting up with our many suggestions for improvement.

pb82 commented:
You're welcome, thanks for merging :-) @jazzyb: most of your suggestions have been implemented in the last commits. Also looking forward to your sqlite_ecto adapter, always handy to have serverless database support!
